MiniApp/Classes/MiniAppDownloader.swift

```diff
@@ -23,27 +23,56 @@ class MiniAppDownloader {
-        if miniAppStatus.isDownloaded(appId: appId, versionId: versionId) {
+        if !isMiniAppAlreadyDownloaded(appId: appId, versionId: versionId) {
+            self.manifestDownloader.fetchManifest(apiClient: self.miniAppClient, appId: appId, versionId: versionId) { (result) in
+                 switch result {
+                 case .success(let responseData):
+                     self.downloadMiniApp(urls: responseData.manifest, to: miniAppStoragePath) { downloadResult in
+                         switch downloadResult {
+                         case .success:
+                             self.miniAppStorage.cleanVersions(for: appId, differentFrom: versionId, status: self.miniAppStatus)
+
+                             fallthrough
+                         default:
+                             completionHandler(downloadResult)
+                         }
+                     }
+                 case .failure(let error):
+                     let downloadError = error as NSError
+                     if downloadError.code == NSURLErrorNotConnectedToInternet {
+                         self.getCachedMiniApp(appId: appId, downloadError: downloadError, completionHandler: completionHandler)
+                         return
+                     }
+                     return completionHandler(.failure(error))
+                 }
+             }
+        } else {
             completionHandler(.success(miniAppStoragePath))
-            return
         }
-        self.manifestDownloader.fetchManifest(apiClient: self.miniAppClient, appId: appId, versionId: versionId) { (result) in
-            switch result {
-            case .success(let responseData):
-                self.downloadMiniApp(urls: responseData.manifest, to: miniAppStoragePath) { downloadResult in
-                    switch downloadResult {
-                    case .success:
-                        self.miniAppStorage.cleanVersions(for: appId, differentFrom: versionId, status: self.miniAppStatus)
+    }
 
-                        fallthrough
-                    default:
-                        completionHandler(downloadResult)
-                    }
-                }
-            case .failure(let error):
-                return completionHandler(.failure(error))
+    func isMiniAppAlreadyDownloaded(appId: String, versionId: String) -> Bool {
+        if miniAppStatus.isDownloaded(appId: appId, versionId: versionId) {
+            let versionDirectory = FileManager.getMiniAppVersionDirectory(with: appId, and: versionId)
+            var isDirectory: ObjCBool = true
+            if FileManager.default.fileExists(atPath: versionDirectory.path, isDirectory: &isDirectory) {
+                return true
             }
         }
+        return false
+    }
+
+    func getCachedMiniApp(appId: String, downloadError: NSError, completionHandler: @escaping (Result<URL, Error>) -> Void) {
```

Comment:
It looks like there isn't a reason that this function needs to be passed a completionHandler. It's not doing anything asynchronous, right? 

I think it would be easier to follow if you make `getCachedMiniApp` return `URL?`. And then in the `download` function, you just need to check if it returned a null value. Then you don't need to pass the `downloadError` or `completionHandler` to `getCachedMiniApp`.

Reply:
Done 👍

Reply:
That's a better approach, updated code now 👍